airbyte-cdk/python/airbyte_cdk/sources/streams/call_rate.py

```diff
@@ -0,0 +1,445 @@
+            self._calls_num = self._call_limit - available_calls
+
+        if call_reset_ts is not None and call_reset_ts != self.current_window.end:
+            logger.info("got update from api, adjusting reset time from %s to %s", self.current_window.end, call_reset_ts)
+            self.current_window.end = call_reset_ts
+
+    def _update_current_window(self) -> None:
+        now = datetime.datetime.now()
+        if now > self.current_window.end:
+            logger.info("current window %s has being reset, %s calls available now", self.current_window, self._call_limit)
+            self._current_window = TimeWindow(self._current_window.end, self._current_window.end + self._offset)
+            self._calls_num = 0
+
+
+class MovingWindowCallRatePolicy(AbstractCallRatePolicy):
+    """
+    Policy to control requests rate implemented on top of PyRateLimiter lib.
+    """
+
+    def __init__(self, rates: list[Rate]):
+        """Constructor
+
+        :param rates: list of rates, the order is important and must be ascending
+        """
+        pyrate_rates = [PyRateRate(limit=rate.limit, interval=int(rate.interval.total_seconds() * 1000)) for rate in rates]
+        self._bucket = InMemoryBucket(pyrate_rates)
+        # Limiter will create the background task that clears old requests in the bucket
+        self._limiter = Limiter(self._bucket)
+
+    def try_acquire(self, request: Any, weight: int) -> None:
+        try:
+            self._limiter.try_acquire(request, weight=weight)
+        except BucketFullException as exc:
+            item = self._limiter.bucket_factory.wrap_item(request, weight)
+            assert isinstance(item, RateItem)
+
+            with self._limiter.lock:
+                time_to_wait = self._bucket.waiting(item)
+                assert isinstance(time_to_wait, int)
+
+                raise CallRateLimitHit(
+                    error=str(exc.meta_info["error"]),
+                    item=request,
+                    weight=int(exc.meta_info["weight"]),
+                    rate=str(exc.meta_info["rate"]),
+                    time_to_wait=timedelta(milliseconds=time_to_wait),
+                )
+
+    def update(self, available_calls: Optional[int], call_reset_ts: Optional[datetime.datetime]) -> None:
+        """Adjust call bucket to reflect the state of the API server
+
+        :param available_calls:
+        :param call_reset_ts:
+        :return:
+        """
+        if available_calls is not None and call_reset_ts is None:  # we do our best to sync buckets with API
+            if available_calls == 0:
+                with self._limiter.lock:
+                    items_to_add = self._bucket.count() < self._bucket.rates[0].limit
+                    if items_to_add > 0:
+                        now: int = TimeClock().now()  # type: ignore[no-untyped-call]
+                        self._bucket.put(RateItem(name="dummy", timestamp=now, weight=items_to_add))
```

Comment:
does the name matter?

Reply:
not really, the name to help debug issues and will be shown when the exception is thrown inside py-rate, we replace it with the actual request item when passing it to the calling code